aten/src/ATen/ThreadLocalState.cpp

```diff
@@ -9,7 +9,7 @@
 namespace at {
 
 ThreadLocalState::ThreadLocalState(bool keep_grad_mode)
-    : dispatch_key_(c10::impl::tls_local_dispatch_key_set()),
+    : dispatch_key_(c10::impl::snapshot_tls_keyset()),
```

Comment:
given that you add grad mode later into the new TLS state, should we copy the whole POD state here?

Reply:
That kind of already happens with the `grad_mode_enabled_ = GradMode::is_enabled();` call. The only difference is that the source of dispatch keys and grad mode is the same TLS instead of two different ones. The more I look at this the less thrilled I am with the interaction of all the TLS's. (This PR isn't making it worse, it's just highlighting that we have kind of a lot of duplicated state.)

Reply:
Although that wouldn't respect `keep_grad_mode`.

Reply:
oh, right, re. `keep_grad_mode` - the only use case for this is to avoid enabling grad mode within autograd thread itself, when transitioning from the forward pass into the backward pass